chain-vote/src/shvzk.rs

```diff
@@ -32,30 +34,125 @@ impl ABCD {
 }
 
 /// I, B, A commitments
-#[derive(Clone)]
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
 struct IBA {
     pub i: Commitment,
     pub b: Commitment,
     pub a: Commitment,
 }
 
 // Computed z, w, v
-#[derive(Clone)]
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
 struct ZWV {
     pub z: Scalar,
     pub w: Scalar,
     pub v: Scalar,
 }
 
 /// Proof of unit vector
-#[derive(Clone)]
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
 pub struct Proof {
     ibas: Vec<IBA>,
     ds: Vec<Ciphertext>,
     zwvs: Vec<ZWV>,
     r: Scalar,
 }
 
+impl IBA {
```

Comment:
All serialization from here should be reviewed carefully. For many `chain-vote` structures it is possible to have constant length buffers.